BUG: Categorical doesn't show tzinfo properly

sinhrks opened the pull request:
Closes #10713.
- [x] Check timedelta
- [x] Refactor not to import `DatetimeIndex` and `PeriodIndex` if possible.
- [x] Fix `categories` line break location
- [x] Check `Categorical` `Series` formatting
- [x] Check `Categorical` `DataFrame` formatting 
- [x] Fix `CategoricalIndex` formatting 
## CategoricalIndex problem (befor this PR)

```
import pandas as pd
idx = pd.period_range('2011-01-01 09:00', freq='H', periods=5)
idx = pd.Index(pd.Categorical(idx))
idx
# CategoricalIndex([359409, 359410, 359411, 359412, 359413], categories=[2011-01-01 09:00, 2011-01-01 10:00, 2011-01-01 11:00, 2011-01-01 12:00, 2011-01-01 13:00], ordered=False, dtype='category')
```
## Format differences (after this PR)

I left ordered category representation of `CategoricalIndex` as below because it has `ordered` repr separately.

```
pd.Categorical([1, 2, 3], ordered=True)
# [1, 2, 3]
# Categories (3, int64): [1 < 2 < 3]

pd.Series(pd.Categorical([1, 2, 3], ordered=True))
#0    1
#1    2
#2    3
# dtype: category
# Categories (3, int64): [1 < 2 < 3]

pd.CategoricalIndex([1, 2, 3], ordered=True)
# CategoricalIndex([1, 2, 3], categories=[1, 2, 3], ordered=True, dtype='category')
```

sinhrks committed:
BUG: Categorical doesn't show tzinfo properly

jreback commented on pandas/core/categorical.py:
```diff
@@ -1324,8 +1326,17 @@ def _repr_categories(self):
                     else get_option("display.max_categories"))
+
+        from pandas.tseries.index import DatetimeIndex
+        from pandas.tseries.period import PeriodIndex
```
use ``core.format.format_array`` does all this.

jreback commented on pandas/core/categorical.py:
```diff
@@ -1340,8 +1351,14 @@ def _repr_categories_info(self):
-        levheader = "Categories (%d, %s): " % (len(self.categories),
-                                               self.categories.dtype)
+
+        from pandas.tseries.period import PeriodIndex
```
yeh ``PeriodIndex`` acts slightly differently. what I would do it maybe have another property of ``Index`` , maybe ``.dtype_values`` which returns ``self.dtype``, but for ``PeriodIndex`` returns ``period``. (this will fix some other uses of ``isinstance(...., PeriodIndex)`` type of things (in other parts of the code)

sinhrks replied:
OK, how about using ``inferred_type`` for ``PeriodIndex``, and ``dtype`` for others (for numpy compat).

jreback commented on pandas/core/categorical.py:
```diff
@@ -1054,8 +1054,10 @@ def get_values(self):
-        if isinstance(self.categories, ABCPeriodIndex):
+        from pandas.tseries.index import DatetimeIndex
+        from pandas.tseries.period import PeriodIndex
+        # if we are a datetime and period index, return a string repr
```
use ``core.is_datetimelike``

jreback commented on pandas/core/categorical.py:
```diff
@@ -1054,8 +1054,10 @@ def get_values(self):
+        from pandas.tseries.period import PeriodIndex
+        # if we are a datetime and period index, return a string repr
+        if isinstance(self.categories, (DatetimeIndex, PeriodIndex)):
             return take_1d(np.array(self.categories.to_native_types(), dtype=object),
```
hmm, this should work now as all ``Index`` will give you a str repr

sinhrks commented:
@jreback Thanks, fixed your points and it's ready for review. Checked ``Categorical``, ``Index``, ``Series`` and ``DataFrame`` outputs.

jreback commented on pandas/core/categorical.py:
```diff
@@ -1340,8 +1337,12 @@ def _repr_categories_info(self):
         category_strs = self._repr_categories()
-        levheader = "Categories (%d, %s): " % (len(self.categories),
-                                               self.categories.dtype)
+        if self.categories.inferred_type == 'period':
```
ok, I think to the ``Index`` add a property that returns ``.dtype`` (and returns ``.inferred_type`` for a ``PeriodIndex``), maybe ``getattr(self.categories,'dtype_str',self.categories.dtype)``; this will also allow it to word if ``categories`` are an ``ndarray``, though I don't think we actually allow this.

jreback commented on pandas/core/format.py:
```diff
@@ -207,7 +207,11 @@ def _get_formatted_index(self):
         return fmt_index, have_header
 
     def _get_formatted_values(self):
-        return format_array(self.tr_series.get_values(), None,
```
I think I changed this in another PR I am doing, use do

``return format_array(self.tr_series.values, ...)`` and this wil work

``format_array`` already handles this type of dispatching

jreback commented on pandas/core/format.py:
```diff
@@ -680,8 +684,12 @@ def write(buf, frame, column_format, strcols, longtable=False):
+        if com.is_categorical_dtype(col):
+            col = col.values
         return format_array(
-            (frame.iloc[:, i]).get_values(),
```
same here, just pass ``.values``

jreback commented on pandas/core/format.py:
```diff
@@ -1895,6 +1903,14 @@ def get_formatted_cells(self):
 def format_array(values, formatter, float_format=None, na_rep='NaN',
                  digits=None, space=None, justify='right'):
+
+    if isinstance(values, Index):
```
so here add the test for ``is_categorical_dtype``

and just have another part of the if which handles ``PeriodIndex`` (as in ``com.is_period_arraylike``)), as ``datetime64/timedelta64`` are already dispatched correctly.

and tz's are handled in the ``Datetime64Formatter`` (soon the tz-block formattr will handle as well)

jreback commented on pandas/core/index.py:
```diff
@@ -2994,6 +3003,13 @@ def equals(self, other):
 
+    @property
+    def _formatter_data(self):
+        """
```
this should not be necessary

jreback commented on pandas/core/index.py:
```diff
@@ -471,6 +471,13 @@ def _formatter_func(self):
 
+    @property
+    def _formatter_data(self):
+        """
```
why are you adding this?

jreback commented on pandas/core/index.py:
```diff
@@ -508,24 +515,26 @@ def best_len(values):
             else:
                 return 0
 
+        data = self._formatter_data
```
why is this changed?

sinhrks replied:
@jreback This is because ``Index._formatter_func`` converts categorical to int like below. 


```
i = pd.CategoricalIndex(pd.period_range('2011-01', freq='M', periods=3))
i._formatter_func(i[:2])
# u'[492, 493]'
```

I think we should fix it either by:

- Change passed data (current impl)
- Change ``Index.format_data`` to handle categorical
- Change ``com.pprint_thing`` 

Which is preferable?

jreback replied:
make a nicer ``_formatter_func``, this is what the datetimelikes do, this would override the use of ``com.pprint_thing`` (which is the default), and do a specific formatter

sinhrks replied:
OK, done. I also fixed  ``Categorical.__iter__`` which may return internal repr like below.

```
[ x for x in pd.Categorical(pd.period_range('2011-01', freq='M', periods=3))]
[492, 493, 494]
```

jreback commented on pandas/core/format.py:
```diff
@@ -1895,6 +1895,17 @@ def get_formatted_cells(self):
+    if com.is_categorical_dtype(values):
+        values = values.get_values()
+
+    if isinstance(values, Index):
```
I don't think you need 1902-1907. This is just checking dtype, its up to the formatter to format it. you want to delay converting to the lowest type here (and instead do it in the formatting part which is specific to that type).

jreback commented on pandas/core/format.py:
```diff
@@ -1895,6 +1895,17 @@ def get_formatted_cells(self):
 
 def format_array(values, formatter, float_format=None, na_rep='NaN',
                  digits=None, space=None, justify='right'):
+
```
make a ``CategoricalFormatter`` (a trivial sub-class of ``GenericArray``). Then you can simply override ``_format_strings`` as needed.

sinhrks commented:
Thanks for comments. Added ``CategoricalArrayFormatter`` as ``CategoricalFormatter`` is already occupied, and also added ``PeriodArrayFormatter`` which doesn't exist. Pls check whether it meets your suggestion.

jreback commented on doc/source/whatsnew/v0.17.0.txt:
```diff
@@ -605,6 +605,9 @@ Bug Fixes
 
 - Bug in ``read_stata`` when reading a file with a different order set in ``columns`` (:issue:`10757`)
+- Bug in ``Categorical`` may not represented properly when category contains ``tz`` or ``Period`` (:issue:`10713`)
+- Bug in ``Categorical.__iter__`` may return correct ``datetime`` and ``Period`` (:issue:`10713`)
```
not representing 
not returning

jreback commented on pandas/tseries/period.py:
```diff
@@ -530,6 +532,11 @@ def shift(self, n):
+    @cache_readonly
+    def dtype_str(self):
+        """ return the dtype str of the underlying data """
+        return self.inferred_type
```
when we add a real period dtype then can fix this
can u add a xref to that issue (point to this PR)

jreback commented:
minor comments
looks good
merge when ready
